spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/QueryParser.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2022-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.jpa.repository.query;
+
+import static org.springframework.data.jpa.repository.query.QueryParsingToken.*;
+
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.jpa.domain.JpaSort;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * Operations needed to parse a JPA query.
+ *
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+abstract class QueryParser {
+
+	private final DeclaredQuery declaredQuery;
+
+	QueryParser(DeclaredQuery declaredQuery) {
+		this.declaredQuery = declaredQuery;
+	}
+
+	QueryParser(String query) {
+		this(DeclaredQuery.of(query, false));
+	}
+
+	DeclaredQuery getDeclaredQuery() {
+		return declaredQuery;
+	}
+
+	String getQuery() {
+		return getDeclaredQuery().getQueryString();
+	}
+
+	/**
+	 * Parse the JPA query using its corresponding ANTLR parser.
+	 */
+	abstract ParserRuleContext parse();
+
+	/**
+	 * Generate a query using the original query with an @literal order by} clause added (or amended) based upon the
+	 * provider {@link Sort} parameter.
+	 *
+	 * @param parsedQuery
+	 * @param sort can be {@literal null}
+	 */
+	String createQuery(ParserRuleContext parsedQuery, Sort sort) {
```

Comment:
The is because invoking the ANTLR-generate `visit()` method requires passing in one of the generated context objects. ALL of their context classes implement `ParserRuleContext`, so this was my way to make HQL and JPQL both have a common API.

Basically:

```java
HqlLexer lexer = new HqlLexer(CharStreams.fromString(query));
HqlParser parser = new HqlParser(new CommonTokenStream(lexer));

parser.addErrorListener(new QueryParsingSyntaxErrorListener());

return parser.start();
```
...is the ANTLR pattern to parse a query and get handed back an AST.

```java
new HqlQueryTransformer(sort).visit(parsedQuery)
```
...is the ANTLR way to then apply one of their visitors to that AST. I'm trying to encapsulate this flow inside of `QueryParser`.

The reasons I have this "visible" is so that `QueryEnhancer` can apply various strategies based on what function it's doing. For instance, when you do `applySorting`, it's supposed to raise an `IllegalArgumentException`. 

```java
@Override
public String applySorting(Sort sort) {

	try {
		ParserRuleContext parsedQuery = queryParser.parse();

		if (parsedQuery == null) {
			return "";
		}

		return queryParser.createQuery(parsedQuery, sort);
	} catch (QueryParsingSyntaxError e) {
		LOG.warn(e);
		throw new IllegalArgumentException(e);
	}
}
```

However, if you run `detectAlias`, it's does NOT throw anything, but instead will simply return `null` for an invalid query.

```java
@Override
public String detectAlias() {

	try {
		ParserRuleContext parsedQuery = queryParser.parse();

		if (parsedQuery == null) {

			LOG.warn("Failed to parse " + queryParser.getQuery() + ". See console for more details.");
			return null;
		}

		return queryParser.findAlias(parsedQuery);
	} catch (QueryParsingSyntaxError e) {
		LOG.warn(e);
		return null;
	}
}
```

This is where it would be nice if we had a chance to refactor `QueryEnhancer`. But that feels like too big a change. I figured if we kept all the ANTLR specifics inside `QueryParsingEnhancer`, we could minimize that leakage.